sklearn/linear_model/ridge.py

```diff
@@ -367,11 +367,29 @@ def _ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
                       return_n_iter=False, return_intercept=False,
                       X_scale=None, X_offset=None):
 
-    if return_intercept and sparse.issparse(X) and solver != 'sag':
-        if solver != 'auto':
-            warnings.warn("In Ridge, only 'sag' solver can currently fit the "
-                          "intercept when X is sparse. Solver has been "
-                          "automatically changed into 'sag'.")
+    has_sw = sample_weight is not None
+
+    def _select_auto_mode():
+        if return_intercept:
+            # only sag supports fitting intercept directly
+            return "sag"
```

Comment:
This logic duplicates that of line 390.

Reply:
This line executes when argument solver is set to 'auto', line 390 executes when user sets the argument to any other value (except 'sag').  We need to make difference between these two cases, because in one we generate a warning whereas in the other we don't.